### ISSUE TITLE
[Accepted with Revisions] SDL 0081 - SDLInterfaceManager

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#236]
Hello SDL community,

The review of "SDLInterfaceManager" begins now and runs through August 8, 2017. The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0081-SDLInterfaceManager.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/236

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by joeljfischer** [comment]
There appear to be several issues with this proposal.

1. `SDLInterfaceManager` is not a good name for this, it's not descriptive enough in my opinion. Perhaps something like `SDLWindowHapticsLocator`.

2. I know that this is example code, but if it's meant to demonstrate that this kind of system is possible, I think it fails in a few key areas:
    * This seems to rely on using `preferredFocusedView` which was introduced in iOS 9 (so this proposal will not work for apps that define a lower minimum version, which the proposal does not indicate) and deprecated in iOS 10 (and so may break in the future). The preferred alternative according to Apple is `preferredFocusEnvironments`, and it's not clear how this would be used to provide the same functionality.
    * This proposal relies on developers building their projection using native iOS types, but this has problems:
       * Developers using OpenGL are left out
       * It is impossible to cause touch events just by knowing coordinates. So if the haptic engine says "press the button in this rectangle", it's not clear how we would translate that into running the code associated with the button. We would have to build some sort of drawing manager (effectively a replacement for UIKit) and have developers custom draw elements, or else build a whole bindings framework and force developers to use that when building their UIs, which may not be feasible with native UI elements (e.g. when building a tableview).

3. It's not quite clear how the drawing of focus rects is intended to be done. I had been assuming it would be done on the head unit side, not the projection app side, but the example code shows it done on the app side.

4. `SDLCarWindow` is referenced, but it's unclear what this is.

5. It's unclear how this adapts to and works with animation or view hierarchies. Is UINavigationController used? How are screens navigated if not? How and when are focus rects invalidated as a screen changes? Is this automatic?

6. How do we enforce driver distraction? Do we set required tap target sizes? How about with tableview? With a navigation controller? Other native UI elements? If we design our own UI hierarchy to make this work, how should we design them? How much customizability? What sizes?

There are far too many unanswered questions and this is far too large of a proposal with too little information. I strongly recommend that this proposal is rejected.

**Comment by Toyota-BSalahat** [comment]
> SDLInterfaceManager is not a good name for this, it's not descriptive enough in my opinion. Perhaps something like SDLWindowHapticsLocator.

Agreed. Naming things is hard.

> This seems to rely on using preferredFocusedView which was introduced in iOS 9 (so this proposal will not work for apps that define a lower minimum version, which the proposal does not indicate) and deprecated in iOS 10 (and so may break in the future). The preferred alternative according to Apple is preferredFocusEnvironments, and it's not clear how this would be used to provide the same functionality.

Agreed, I think this is noted in the proposal somewhat. The alternative would be keying off the accessibility focus flags instead, which should be available since iOS 7 or 8 (at least until we are comfortable moving the minimum support version to 9).

> Developers using OpenGL are left out

Agreed. This class would need to allow an interface for the app developer to manually pass us the data required by the spatial struct (x, y, length, width), or we could supply an adapter class to work with openGL. I think that could become work intensive though over time.

> It is impossible to cause touch events just by knowing coordinates. So if the haptic engine says "press the button in this rectangle", it's not clear how we would translate that into running the code associated with the button. We would have to build some sort of drawing manager (effectively a replacement for UIKit) and have developers custom draw elements, or else build a whole bindings framework and force developers to use that when building their UIs, which may not be feasible with native UI elements (e.g. when building a tableview).

I'm not sure I understand this concern. This sounds like a limitation that would exist with the current VPM implementation. Apps would receive touch events from SDL the same way they do now. Do I misunderstand?

> It's not quite clear how the drawing of focus rects is intended to be done. I had been assuming it would be done on the head unit side, not the projection app side, but the example code shows it done on the app side.

I think I commented this in the code, but this is my fault for not being clear. That code is for debugging purposes only so that any interested SDLC member can see how this example identifies views in any UIKit app. It would not go in the proxy.

> SDLCarWindow is referenced, but it's unclear what this is.

SDLCarWindow is an abstraction that has been developed to support VPM. It is something I believe will be proposed in the near future. It adds an interface to the proxy to allow app developers to draw off screen view controllers that are mirrored to the HU. It is out of the scope of the haptic features, but this proposal has been designed to be able to plug and play with any UIWindow based solution.

> It's unclear how this adapts to and works with animation or view hierarchies. Is UINavigationController used? How are screens navigated if not? How and when are focus rects invalidated as a screen changes? Is this automatic?

Good questions. This should absolutely work with UINavigationController. I did not notice any issues with that in my testing. The code crawls the entire view hierarchy starting at the root view controller. Any container that does not break the view tree should be supported (any UIKit container)

This proposal does demonstrate code that automatically updates the focus rects any time a view appears on screen, however reacting to events like scrolling and view updates would most likely need to be called from the app. Much like you would call reloadData on a tableview, the app would need to call reloadFocus (or something similar) to refresh the focus rects. We could make some more of this automatic, but I am not sure we will ever be able to catch every scenario.

>How do we enforce driver distraction? Do we set required tap target sizes? How about with tableview? With a navigation controller? Other native UI elements? If we design our own UI hierarchy to make this work, how should we design them? How much customizability? What sizes?

This is a problem with video projection apps, not with the haptic interface. It is outside of the scope of this proposal.

**Comment by joeljfischer** [comment]
> > This seems to rely on using preferredFocusedView which was introduced in iOS 9 (so this proposal will not work for apps that define a lower minimum version, which the proposal does not indicate) and deprecated in iOS 10 (and so may break in the future). The preferred alternative according to Apple is preferredFocusEnvironments, and it's not clear how this would be used to provide the same functionality.

> Agreed, I think this is noted in the proposal somewhat. The alternative would be keying off the accessibility focus flags instead, which should be available since iOS 7 or 8 (at least until we are comfortable moving the minimum support version to 9).

Okay, but then the proposal shouldn't be accepted as written, and using accessibility may have similar issues. Out of curiosity, what is the downside of just traversing the view hierarchy? Why deal with this focus engine or accessibility stuff at all?

> > Developers using OpenGL are left out

> Agreed. This class would need to allow an interface for the app developer to manually pass us the data required by the spatial struct (x, y, length, width), or we could supply an adapter class to work with openGL. I think that could become work intensive though over time.

Since most / all current developers are using OpenGL this seems like a major downside.

> > It is impossible to cause touch events just by knowing coordinates. So if the haptic engine says "press the button in this rectangle", it's not clear how we would translate that into running the code associated with the button. We would have to build some sort of drawing manager (effectively a replacement for UIKit) and have developers custom draw elements, or else build a whole bindings framework and force developers to use that when building their UIs, which may not be feasible with native UI elements (e.g. when building a tableview).

> I'm not sure I understand this concern. This sounds like a limitation that would exist with the current VPM implementation. Apps would receive touch events from SDL the same way they do now. Do I misunderstand?

You're correct in that this would be a concern of VPM apps in general *for those that implement native UI views*. We currently only support piping, but as a dependency of this interface is to have built in support of UIKit based video streaming, and any of those will have this problem. OpenGL would be easier for developers to deal with. Since this proposal only affects UIKit based views, it has these issues.

> > SDLCarWindow is referenced, but it's unclear what this is.

> SDLCarWindow is an abstraction that has been developed to support VPM. It is something I believe will be proposed in the near future. It adds an interface to the proxy to allow app developers to draw off screen view controllers that are mirrored to the HU. It is out of the scope of the haptic features, but this proposal has been designed to be able to plug and play with any UIWindow based solution.

If this is a dependency, (or any UIWindow based solution is a dependency) then that would need to be discussed and accepted first (or simultaneously if the timing / need is dire), in my opinion.

> > It's unclear how this adapts to and works with animation or view hierarchies. Is UINavigationController used? How are screens navigated if not? How and when are focus rects invalidated as a screen changes? Is this automatic?

> Good questions. This should absolutely work with UINavigationController. I did not notice any issues with that in my testing. The code crawls the entire view hierarchy starting at the root view controller. Any container that does not break the view tree should be supported (any UIKit container)

> This proposal does demonstrate code that automatically updates the focus rects any time a view appears on screen, however reacting to events like scrolling and view updates would most likely need to be called from the app. Much like you would call reloadData on a tableview, the app would need to call reloadFocus (or something similar) to refresh the focus rects. We could make some more of this automatic, but I am not sure we will ever be able to catch every scenario.

> > How do we enforce driver distraction? Do we set required tap target sizes? How about with tableview? With a navigation controller? Other native UI elements? If we design our own UI hierarchy to make this work, how should we design them? How much customizability? What sizes?

> This is a problem with video projection apps, not with the haptic interface. It is outside of the scope of this proposal.

I'm responding to these both in one. Because this is the first proposal that puts forward the use of native UIKit UI in video projection apps and natively supported by the SDL iOS library, I feel this is exactly the conversation we need to be having. If you don't agree, then I would argue that this proposal is being proposed too early, and a proposal explicitly for the purpose of discussing the promotion and native support of UIKit based UIs for projection apps needs to be proposed and discussed before we can discuss this one.

When you say UINavigationController / UITabBarController work, I understand that they likely *would* work, the question is *should* they, and should we be promoting their use. First, the UINavigation / UITabBar controllers likely will not make for car friendly UIs. Second, We have no way to cause button presses on these native UIKit objects without modification. Again, this all adds up to significantly more work than is laid out in this proposal.

**Comment by Toyota-BSalahat** [comment]
> Okay, but then the proposal shouldn't be accepted as written, and using accessibility may have similar issues. Out of curiosity, what is the downside of just traversing the view hierarchy? Why deal with this focus engine or accessibility stuff at all?

This is listed as an alternative approach in the proposal, as written. Is this not the point of that section? The SDLC will need to discuss the potential alternatives and decide the best approach.

How will you identify which controls an app deems worth having focusable? Do we want to focus labels? Do we want to focus disabled or hidden buttons? Do we want to be responsible for dictating that logic when Apple has provided interfaces for apps to specify this that are standard? Furthermore, if the SDLC does in the future decide to support app-side highlighting, would we not want to emulate the familiar-to-app-developers interface that the feedback engine provides? I suggest mimicking the interfaces that Apple provides and utilizing them when it makes sense, but technically yes we could take the burden on of making these determinations with the raw view data.

> Since most / all current developers are using OpenGL this seems like a major downside.

There is nothing to stop us from working with an OpenGL engineer in the future and architecting better support for this. This proposal is aimed at adding support to UIKit though, and basic generic support. I would think that would be best as a separate proposal anyways, since the involved logic would likely be quite different.

> I'm responding to these both in one. Because this is the first proposal that puts forward the use of native UIKit UI in video projection apps and natively supported by the SDL iOS library, I feel this is exactly the conversation we need to be having. If you don't agree, then I would argue that this proposal is being proposed too early, and a proposal explicitly for the purpose of discussing the promotion and native support of UIKit based UIs for projection apps needs to be proposed and discussed before we can discuss this one.

Your concerns are not invalid, but there is nothing stopping a VPM app from projecting native views at this time. This can be done using the current video projection APIs right now by an app partner. This proposal adds a class to provide an interface to the OEM haptic engine. It supports UIKit because UIKit is easily supported, but it does not change the current nature of video projection. These things should be discussed and standardized by the SDLC, but that is not the aim of this proposal, nor do I think it belongs in this proposal. No additional risks are added by projecting UIKit UIs. The same problems can be created in OpenGL or any other graphics engine just as easily.


> When you say UINavigationController / UITabBarController work, I understand that they likely would work, the question is should they, and should we be promoting their use. First, the UINavigation / UITabBar controllers likely will not make for car friendly UIs.

Again, there is nothing from stopping an app from doing this currently. All of the problems will exist with video project as it is today. If an app developer wants to subclass UITabBarController or UINavigationController to make it visually car friendly, I am not sure that we should stop them. Regardless, I don't think this problem specifically relates to the haptic engine. The haptic engine should not be more limited than the VPM feature is.

> Second, We have no way to cause button presses on these native UIKit objects without modification. Again, this all adds up to significantly more work than is laid out in this proposal.

SDL reports the touch coordinates to the app. I would think that the app could handle the touch even with a UIKit based UI the same as it could today. If that is a limitation of SDL, that fact does not change by supporting the OEM haptic engine.

> If this is a dependency, (or any UIWindow based solution is a dependency) then that would need to be discussed and accepted first (or simultaneously if the timing / need is dire), in my opinion

It is not a dependency, this proposal will simply be compatible with the changes. Since I am aware it is in development and will be proposed soon, I felt it was good to point out.

This proposal does not aim to limit haptic support to UIKit apps. It aims to introduce a manager class for haptic support. We can, and should continue to support the manager to continue to make integration as painless as possible for all types of VPM apps. In an effort to keep the scope small this proposal aims to add the base manager pass-through for the new RPC, and UIKit support.

**Comment by joeljfischer** [comment]
> This is listed as an alternative approach in the proposal, as written. Is this not the point of that section? The SDLC will need to discuss the potential alternatives and decide the best approach.

Largely, no, it's not the point of the section. We have used it (incorrectly in my mind) that way in the past, but that section is designed to be "considered alternatives that were rejected for these reasons". If we actually want to use an alternative, the proposal should be rewritten so that we can understand the pros and cons of the alternative. There's certainly not enough information on the alternative here to understand the merits.

> How will you identify which controls an app deems worth having focusable? Do we want to focus labels? Do we want to focus disabled or hidden buttons? Do we want to be responsible for dictating that logic when Apple has provided interfaces for apps to specify this that are standard? Furthermore, if the SDLC does in the future decide to support app-side highlighting, would we not want to emulate the familiar-to-app-developers interface that the feedback engine provides? I suggest mimicking the interfaces that Apple provides and utilizing them when it makes sense, but technically yes we could take the burden on of making these determinations with the raw view data.

I don't know how we should do it. You're right that raw view checking isn't ideal. As I've noted, I think the only way to do this in a way that would work is to implement our own UI framework, which is an enormous amount of work. That or you should investigate the `preferredFocusEnvironments`. I don't have a good answer for you, but the current proposal is inadequate here in my opinion.

> Your concerns are not invalid, but there is nothing stopping a VPM app from projecting native views at this time. This can be done using the current video projection APIs right now by an app partner. This proposal adds a class to provide an interface to the OEM haptic engine. It supports UIKit because UIKit is easily supported, but it does not change the current nature of video projection. These things should be discussed and standardized by the SDLC, but that is not the aim of this proposal, nor do I think it belongs in this proposal. No additional risks are added by projecting UIKit UIs. The same problems can be created in OpenGL or any other graphics engine just as easily.

I'm sorry, but I must wholeheartedly disagree. There's an enormous difference between "developers can use UIKit, but it won't work" and "we built a framework for you to use haptic in your UIKit projection app, and it still doesn't work". The former isn't our fault if developers tread that path, the latter most certainly is. And my concerns about UIKit being a wholly poor solution for this have not been addressed. Again, for emphasis, UIKit does not work with projection as it stands today.

> Again, there is nothing from stopping an app from doing this currently. All of the problems will exist with video project as it is today. If an app developer wants to subclass UITabBarController or UINavigationController to make it visually car friendly, I am not sure that we should stop them. Regardless, I don't think this problem specifically relates to the haptic engine. The haptic engine should not be more limited than the VPM feature is.

I haven't seen any evidence that it's possible to subclass it to make it visually friendly in the car, nor adhere to driver distraction guidelines. More importantly, this problem absolutely relates to the haptic engine, because not only do the developer's view controllers have to be haptic friendly, but the native system UI does too. If the user can't choose a new tab or navigation bar items, the UX is useless.

> SDL reports the touch coordinates to the app. I would think that the app could handle the touch even with a UIKit based UI the same as it could today. If that is a limitation of SDL, that fact does not change by supporting the OEM haptic engine.

You're incorrect here. They can't handle the touch event the same as they can if the user is using it on the phone. This is the whole point I've been trying to make. A touch coordinate is not sufficient for the developer, or for SDL, to cause a touch event on the object. We *cannot simulate UIKit touches based on coordinates*. If this is a limitation of SDL, which it is, then the OEM haptic engine is useless for UIKit based apps, meaning that this proposal won't work and should be rejected.

> It is not a dependency, this proposal will simply be compatible with the changes. Since I am aware it is in development and will be proposed soon, I felt it was good to point out.

Again, I must very strongly disagree. This proposal is absolutely dependent on having a working UIKit based projection system. We don't have it, so this proposal should not be considered until that point.

> This proposal does not aim to limit haptic support to UIKit apps. It aims to introduce a manager class for haptic support. We can, and should continue to support the manager to continue to make integration as painless as possible for all types of VPM apps. In an effort to keep the scope small this proposal aims to add the base manager pass-through for the new RPC, and UIKit support.

The whole proposal is strictly about UIKit based haptic support, so I'm very confused about why you say it's not limited to haptic support for UIKit apps. The proposal explicitly states that OpenGL based apps aren't supported by this interface. You contradict yourself when you say, "This proposal does not aim to limit haptic support to UIKit apps" and "In an effort to keep the scope small this proposal aims to add the base manager pass-through for the new RPC, and UIKit support." I'm not sure what "base manager pass-through" means, but the developer can send and receive those RPCs themselves as of v4.7, and the UIKit is an integral part to this proposal.

**Comment by Toyota-BSalahat** [comment]
@davidswi, @Toyota-Sbetts, @joeljfischer  

> Largely, no, it's not the point of the section. We have used it (incorrectly in my mind) that way in the past, but that section is designed to be "considered alternatives that were rejected for these reasons". If we actually want to use an alternative, the proposal should be rewritten so that we can understand the pros and cons of the alternative. There's certainly not enough information on the alternative here to understand the merits.

It seems like this needs to be clarified then in the process, because that is how it has been used up to this date. I do question the wisdom of invalidating a proposal like this though. Many problems have multiple solutions, and the point of the SDLC is to select the best solution. A proposal should present all solutions known to the author so that the SDLC may make that decision. I am not sure what information you feel is missing in this case, as it is simply use of the accessibility object flag, vs the iOS 9+ focusable flag. 

> I don't know how we should do it. You're right that raw view checking isn't ideal. As I've noted, I think the only way to do this in a way that would work is to implement our own UI framework, which is an enormous amount of work. That or you should investigate the preferredFocusEnvironments. I don't have a good answer for you, but the current proposal is inadequate here in my opinion.

I think we will just have to agree to disagree on your point about creating our own UI framework. Apple provides at least 2 properties that serve for an adequate point of reference to determine which views we care about.

As far as preferredFocusEnviroments:
1. This is a trivial change from the current use of preferredFocusedView. It allows you to implement a delegate callback that returns multiple views in a preferred focus order, instead of just the view that should be focused first. https://stackoverflow.com/questions/39479180/tvos-10-preferredfocusview-is-deprecated-how-do-i-change-to-preferredfocusenvi
2. Use of preferredFocusedView was documented in the proposal as 

> If the window's current view is setting a preferred focus view (ie, the view that should be initially focused as per the tvOS spec) it is moved to the front of the parsed focusable views list. This could be used by the OEM haptic engine, if so desired.

It is a nice to have. If you feel it is a major sticking point, it can be removed entirely. However I think that the app being able to specify an array of the order they prefer views to be focused in is actually a major plus side for the haptic engine.

> I'm sorry, but I must wholeheartedly disagree. There's an enormous difference between "developers can use UIKit, but it won't work" and "we built a framework for you to use haptic in your UIKit projection app, and it still doesn't work". The former isn't our fault if developers tread that path, the latter most certainly is. And my concerns about UIKit being a wholly poor solution for this have not been addressed. Again, for emphasis, UIKit does not work with projection as it stands today.

Again, I suppose we need to agree to disagree. UIKit is a tool to create a user interface. It is neither better or worse equipped to create a projection UI than any alternative. Using views and layers you can accomplish any interface you could design in OpenGL. I also disagree that UIKit can't be used today. There is absolutely nothing stopping a developer from creating a UIView and projecting it to SDL via VPM. It is trivial to scrape an image from the device and project through the current interface.


> I haven't seen any evidence that it's possible to subclass it to make it visually friendly in the car, nor adhere to driver distraction guidelines. More importantly, this problem absolutely relates to the haptic engine, because not only do the developer's view controllers have to be haptic friendly, but the native system UI does too. If the user can't choose a new tab or navigation bar items, the UX is useless.

You haven't seen evidence that an app developer can subclass and customize the appearance of a UIKit view? Are you sure? The interface proposed supports these objects. The user would absolutely be able to select them, the UX is not useless.

> You're incorrect here. They can't handle the touch event the same as they can if the user is using it on the phone. This is the whole point I've been trying to make. A touch coordinate is not sufficient for the developer, or for SDL, to cause a touch event on the object. We cannot simulate UIKit touches based on coordinates. If this is a limitation of SDL, which it is, then the OEM haptic engine is useless for UIKit based apps, meaning that this proposal won't work and should be rejected.

You are somewhat incorrect here. You are correct that there is no production app quality way to simulate a UITouch event, but It is absolutely possible in iOS to map a touch event to a view. UIView provides many functions to do this, and the entire responder chain that iOS is built on is based off of those functions. This could even be handled inside of this manager class if its a better integration experience. The class could return the selected view any time a view is selected on the head unit, but I think it is better to let the app handle this. It is not a particularly unusual thing to do in iOS.  The hitTest method on UIView is frequently used to this end. 

And again, this would exist with a VPM app today, if it decided to use UIKit.

> The whole proposal is strictly about UIKit based haptic support, so I'm very confused about why you say it's not limited to haptic support for UIKit apps. The proposal explicitly states that OpenGL based apps aren't supported by this interface. You contradict yourself when you say, "This proposal does not aim to limit haptic support to UIKit apps" and "In an effort to keep the scope small this proposal aims to add the base manager pass-through for the new RPC, and UIKit support." I'm not sure what "base manager pass-through" means, but the developer can send and receive those RPCs themselves as of v4.7, and the UIKit is an integral part to this proposal.

> Apps that do not use UIKit will have to either set the focusableViews array manually, call the haptic RPC manually, or the InterfaceManager will need to allow apps to pass in an array of the data needed for the InterfaceManager to build out the spatial structures.

**Comment by mfoedisch** [comment]
I have a high-level clarification question on the approach. Via the SendHapticData RPC an app will provide info about the location of it's focusable and selectable items. In case of the user selecting an item, a OnTouchEvent will be sent to the app (is that correct?). But how would the app be informed that a change of focus happened and the app needs to highlight another item?

Is that described anywhere explicitly?

**Comment by joeljfischer** [comment]
@mfoedisch The app highlighting was done for debug purposes currently, the head unit is currently expected to be doing the highlighting.

**Comment by Toyota-BSalahat** [comment]
@mfoedisch, @joeljfischer is correct. The in-app highlighting in this example is just for demonstration in the even any interested party wants to play with the algorithm.

App side focusing was not included in this proposal. We have discussed this internally and we felt that it would be better to consider that feature in the future, and probably use an additional RPC for that type of communication. One of the reasons that @davidswi thought to model the manager around the focus engine is so the manager could be extended in the future to support app-side highlighting, if that is a route that we decide to go in the future.

**Comment by mfoedisch** [comment]
@Toyota-BSalahat , @joeljfischer : thanks, this clarifies!

**Comment by theresalech** [comment]
@Toyota-BSalahat please advise when a new PR has been entered to update the proposal to reflect the agreed upon revisions.  I'll then merge the PR so the proposal is up to date, and enter issues in respective repositories for implementation. Thanks!

**Comment by theresalech** [comment]
The Steering Committee has agreed to accept this proposal with revisions, requesting that the manager use a protocol-oriented design pattern.  The Steering Committee will also have an additional workshop to discuss a larger video projection mode feature.

**Comment by theresalech** [comment]
PR including revisions has been merged. 

Issue has been entered: [[SDL 0081] SDLInterfaceManager](https://github.com/smartdevicelink/sdl_ios/issues/698)
